File: Sources/NockerlDesign/Components/NockerlStepper.swift
// NockerlStepper: process steps for wizards/flows (stepper.mdx).
//
// 32pt discs on the pill silhouette:
//   - DONE:      solid `accentPrimary` + checkmark knockout. **CYAN is the terminal /
//                progress seal (B14 ratified: flow completion is brand progress; green
//                stays for validation messages)**. No green anywhere here.
//   - CURRENT:   solid accent + the step number (reduce-fills, 2026-07-13: the
//                `accentPrimarySoft` halo wash is DROPPED; the accent disc + cyan
//                connectors carry "which step is current", state by indicator not fill).
//   - UPCOMING:  a recessed `canvasAlt` well with a `divider` border, muted number.
//   - ERROR:     solid `statusError` + ✕ knockout (pair with visible error text, law §13).
//
// Connectors: 3pt lines on the squared track radius (`divider` un-reached, solid
// accent completed). Horizontal (labels below) + vertical (labels beside, 360pt
// footprint = size.container.lg). State semantics are ONE pure function shared with
// Compose (`nockerlStepVisualState`), contract-tested on this rail.

import SwiftUI

/// One step of a ``NockerlStepper``.
public struct NockerlStep {
    /// The step label.
    public let label: String
    /// Optional supporting line (vertical orientation only).
    public let description: String?

    /// Create a step.
    public init(_ label: String, description: String? = nil) {
        self.label = label
        self.description = description
    }
}

/// Stepper layout axis.
public enum NockerlStepperOrientation: Equatable {
    /// Discs in a row, labels below.
    case horizontal
    /// Discs in a left rail, labels beside. Clamps to the 360pt footprint.
    case vertical
}

/// The visual state of one step (shared semantics with Compose).
enum NockerlStepVisualState: Equatable {
    case done, current, upcoming, error
}

/// Resolve a step's visual state, the identical rule on both platforms.
func nockerlStepVisualState(index: Int, current: Int, errorAt: Int?) -> NockerlStepVisualState {
    if index == current, errorAt == current { return .error }
    if index < current { return .done }
    if index == current { return .current }
    return .upcoming
}

/// Process steps with the ratified cyan progress discs.
public struct NockerlStepper: View {
    private let steps: [NockerlStep]
    private let current: Int
    private let orientation: NockerlStepperOrientation
    private let errorAt: Int?
    private let onStepClick: ((Int) -> Void)?

    /// Geometry (the ratified recipe).
    static let discSize: CGFloat = NockerlSpace.space8
    static let connectorThickness: CGFloat = 3
    static let ringWidth: CGFloat = NockerlSpace.space1
    static let horizontalColumnWidth: CGFloat = 72
    /// Vertical footprint: the `size.container.lg` token (360).
    static let verticalMaxWidth: CGFloat = 360

    /// Create a stepper.
    /// - Parameters:
    ///   - steps: the steps, in order.
    ///   - current: zero-based active index.
    ///   - orientation: horizontal (default) or vertical.
    ///   - errorAt: step index in the error state (honored when == current).
    ///   - onStepClick: when set, DONE steps become tappable (jump back only).
    public init(
        steps: [NockerlStep],
        current: Int,
        orientation: NockerlStepperOrientation = .horizontal,
        errorAt: Int? = nil,
        onStepClick: ((Int) -> Void)? = nil
    ) {
        self.steps = steps
        self.current = current
        self.orientation = orientation
        self.errorAt = errorAt
        self.onStepClick = onStepClick
    }

    public var body: some View {
        StepperBody(
            steps: steps,
            current: current,
            orientation: orientation,
            errorAt: errorAt,
            onStepClick: onStepClick
        )
    }

    private struct StepperBody: View {
        let steps: [NockerlStep]
        let current: Int
        let orientation: NockerlStepperOrientation
        let errorAt: Int?
        let onStepClick: ((Int) -> Void)?

        @Environment(\.colorScheme) private var colorScheme

        var body: some View {
            let palette = NockerlPalette.resolve(colorScheme)
            switch orientation {
            case .horizontal: horizontal(palette: palette)
            case .vertical: vertical(palette: palette)
            }
        }

        private func horizontal(palette: NockerlPalette) -> some View {
            HStack(alignment: .top, spacing: 0) {
                ForEach(steps.indices, id: \.self) { index in
                    let state = nockerlStepVisualState(index: index, current: current, errorAt: errorAt)
                    if index > 0 {
                        connector(done: index <= current, palette: palette)
                            .frame(minWidth: NockerlSpace.space5)
                            // Center the 3pt line on the disc row (ring inset included).
                            .padding(.top, NockerlStepper.ringWidth + (NockerlStepper.discSize - NockerlStepper.connectorThickness) / 2)
                    }
                    stepColumn(index: index, step: steps[index], state: state, palette: palette)
                }
            }
        }

        private func vertical(palette: NockerlPalette) -> some View {
            VStack(alignment: .leading, spacing: 0) {
                ForEach(steps.indices, id: \.self) { index in
                    let state = nockerlStepVisualState(index: index, current: current, errorAt: errorAt)
                    let isLast = index == steps.count - 1

                    HStack(alignment: .top, spacing: NockerlSpace.space3) {
                        VStack(spacing: NockerlSpace.space1) {
                            disc(index: index, state: state, palette: palette)
                            if !isLast {
                                RoundedRectangle(cornerRadius: NockerlRadius.track, style: .continuous)
                                    .fill(index < current ? palette.accentPrimary : palette.divider)
                                    .frame(width: NockerlStepper.connectorThickness)
                                    .frame(minHeight: NockerlSpace.space5)
                            }
                        }
                        .frame(width: NockerlStepper.discSize + NockerlStepper.ringWidth * 2)

                        VStack(alignment: .leading, spacing: NockerlSpace.space05) {
                            label(steps[index].label, state: state, palette: palette)
                            if let description = steps[index].description {
                                Text(description)
                                    .font(.nockerl(size: NockerlFontSize.size12, weight: .light))
                                    .foregroundColor(palette.onCardMuted)
                            }
                        }
                        .padding(.top, NockerlStepper.ringWidth)
                    }
                    .padding(.bottom, isLast ? 0 : NockerlSpace.space3)
                }
            }
            .frame(maxWidth: NockerlStepper.verticalMaxWidth, alignment: .leading)
        }

        @ViewBuilder
        private func stepColumn(
            index: Int,
            step: NockerlStep,
            state: NockerlStepVisualState,
            palette: NockerlPalette
        ) -> some View {
            let column = VStack(spacing: NockerlSpace.space2) {
                disc(index: index, state: state, palette: palette)
                label(step.label, state: state, palette: palette)
                    .multilineTextAlignment(.center)
            }
            .frame(width: NockerlStepper.horizontalColumnWidth)

            if let onStepClick, state == .done {
                Button {
                    onStepClick(index)
                } label: {
                    column.contentShape(Rectangle())
                }
                .buttonStyle(.plain)
            } else {
                column
            }
        }

        /// The 32pt step disc, per-state per the ratified recipe.
        @ViewBuilder
        private func disc(index: Int, state: NockerlStepVisualState, palette: NockerlPalette) -> some View {
            let core = ZStack {
                switch state {
                case .done, .current, .error:
                    let fill = state == .error ? palette.statusError : palette.accentPrimary
                    Circle().fill(fill)
                    // The shared top catch-light, the same surfaceHighlight token.
                    Circle()
                        .strokeBorder(
                            LinearGradient(
                                colors: [palette.surfaceHighlight, .clear],
                                startPoint: .top,
                                endPoint: .center
                            ),
                            lineWidth: NockerlSpace.spacePx
                        )
                    if state == .current {
                        Text("\(index + 1)")
                            .font(.nockerl(size: NockerlFontSize.size14, weight: .medium))
                            .foregroundColor(palette.onAccent)
                    } else {
                        Image(systemName: state == .done ? "checkmark" : "xmark")
                            .font(.system(size: NockerlFontSize.size12, weight: .medium))
                            .foregroundColor(palette.onAccent)
                    }
                case .upcoming:
                    Circle().fill(palette.canvasAlt)
                    Circle().strokeBorder(palette.divider, lineWidth: NockerlSpace.spacePx)
                    Text("\(index + 1)")
                        .font(.nockerl(size: NockerlFontSize.size14, weight: .medium))
                        .foregroundColor(palette.onCardMuted)
                }
            }
            .frame(width: NockerlStepper.discSize, height: NockerlStepper.discSize)
            .shadow(
                color: state == .upcoming ? .clear : palette.shadowTint.opacity(NockerlCardElevation.level1.tintAlpha),
                radius: NockerlElevation.level1,
                x: 0,
                y: NockerlElevation.level1 / 2
            )
            .accessibilityLabel(accessibleName(for: state, index: index))

            // Reduce-fills (law §6): the CURRENT step reads via its accent DISC + the cyan
            // connectors; the accentPrimarySoft "which step is current" halo wash is dropped
            // (state by the indicator, not a fill). Uniform ring padding keeps discs aligned.
            core.padding(NockerlStepper.ringWidth)
        }

        private func label(_ text: String, state: NockerlStepVisualState, palette: NockerlPalette) -> Text {
            Text(text)
                .font(.nockerl(size: NockerlFontSize.size12, weight: state == .upcoming ? .light : .medium))
                .foregroundColor(
                    state == .error
                        ? palette.statusError
                        : (state == .upcoming ? palette.onCardMuted : palette.onCard)
                )
        }

        private func connector(done: Bool, palette: NockerlPalette) -> some View {
            RoundedRectangle(cornerRadius: NockerlRadius.track, style: .continuous)
                .fill(done ? palette.accentPrimary : palette.divider)
                .frame(height: NockerlStepper.connectorThickness)
                .frame(maxWidth: .infinity)
        }

        private func accessibleName(for state: NockerlStepVisualState, index: Int) -> String {
            let position = "Step \(index + 1) of \(steps.count)"
            switch state {
            case .done: return "\(position), completed"
            case .current: return "\(position), current"
            case .upcoming: return "\(position), upcoming"
            case .error: return "\(position), failed"
            }
        }
    }
}

File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlStepper.kt
package com.dizyx.nockerl.design.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors
import com.dizyx.nockerl.design.tokens.NockerlElevation
import com.dizyx.nockerl.design.tokens.NockerlProgressTrackShape
import com.dizyx.nockerl.design.tokens.nockerlLitSurface
import com.dizyx.nockerl.design.tokens.nockerlRecessedSurface
import com.dizyx.nockerl.design.tokens.nockerlShadow

/**
 * The **stepper**: process steps for wizards/flows (stepper.mdx; the
 * coverage gap closed). Discs at 32dp on the pill silhouette:
 * - **DONE**: solid `accentPrimary` + a check knocked out in `onAccent`, level-1
 *   lift + catch-light. **CYAN is the terminal/progress seal (ratified:
 *   flow completion is brand progress; green stays for validation messages)**.
 *   There is no green anywhere in this component.
 * - **CURRENT**: solid accent + the step number, ringed by an
 *   `accentPrimarySoft` halo (natives express the web's gradient as solid +
 *   catch-light, per the solid-primary doctrine).
 * - **UPCOMING**: a recessed `canvasAlt` well with a `divider` border and a
 *   muted number ("fields sink").
 * - **ERROR** (when [errorAt] == [current]): solid `statusError` + the ✕
 *   knockout; pair it with visible error text at the call site (law §13).
 *
 * Connectors are 3dp lines on the squared track radius: `divider` for
 * un-reached segments, solid accent for completed ones.
 *
 * @param steps the steps, in order.
 * @param current zero-based active index (drives done/current/upcoming).
 * @param modifier outer modifier.
 * @param orientation horizontal (default) or vertical (max 360dp, the
 *   `size.container.lg` footprint).
 * @param errorAt step index in the error state (honored when == [current]).
 * @param onStepClick when set, DONE steps become tappable (jump BACK only;
 *   forward jumps stay gated by the host wizard).
 */
@Composable
fun NockerlStepper(
    steps: List<NockerlStep>,
    current: Int,
    modifier: Modifier = Modifier,
    orientation: NockerlStepperOrientation = NockerlStepperOrientation.HORIZONTAL,
    errorAt: Int? = null,
    onStepClick: ((Int) -> Unit)? = null,
) {
    when (orientation) {
        NockerlStepperOrientation.HORIZONTAL ->
            HorizontalStepper(steps, current, errorAt, onStepClick, modifier)
        NockerlStepperOrientation.VERTICAL ->
            VerticalStepper(steps, current, errorAt, onStepClick, modifier)
    }
}

/** One step of a [NockerlStepper]. */
data class NockerlStep(
    /** The step label. */
    val label: String,
    /** Optional supporting line (vertical orientation only). */
    val description: String? = null,
)

/** Stepper layout axis. */
enum class NockerlStepperOrientation {
    /** Discs in a row, labels below. */
    HORIZONTAL,

    /** Discs in a left rail, labels beside. Clamps to the 360dp footprint. */
    VERTICAL,
}

/** The visual state of one step (shared semantics with Swift; pure + testable). */
internal enum class NockerlStepVisualState { DONE, CURRENT, UPCOMING, ERROR }

/** Resolve a step's visual state (identical rule on both platforms). */
internal fun nockerlStepVisualState(
    index: Int,
    current: Int,
    errorAt: Int?,
): NockerlStepVisualState =
    when {
        index == current && errorAt == current -> NockerlStepVisualState.ERROR
        index < current -> NockerlStepVisualState.DONE
        index == current -> NockerlStepVisualState.CURRENT
        else -> NockerlStepVisualState.UPCOMING
    }

@Composable
private fun HorizontalStepper(
    steps: List<NockerlStep>,
    current: Int,
    errorAt: Int?,
    onStepClick: ((Int) -> Unit)?,
    modifier: Modifier,
) {
    Row(modifier = modifier, verticalAlignment = Alignment.Top) {
        steps.forEachIndexed { index, step ->
            val state = nockerlStepVisualState(index, current, errorAt)
            if (index > 0) {
                Connector(
                    done = index <= current,
                    modifier =
                        Modifier
                            .weight(1f)
                            .widthIn(min = 20.dp)
                            // Center the 3dp line on the 32dp disc row.
                            .padding(top = (STEP_DISC_SIZE - CONNECTOR_THICKNESS) / 2),
                )
            }
            StepColumn(index, step, state, onStepClick)
        }
    }
}

@Composable
private fun VerticalStepper(
    steps: List<NockerlStep>,
    current: Int,
    errorAt: Int?,
    onStepClick: ((Int) -> Unit)?,
    modifier: Modifier,
) {
    val colors = LocalNockerlColors.current
    Column(modifier = modifier.widthIn(max = VERTICAL_MAX_WIDTH)) {
        steps.forEachIndexed { index, step ->
            val state = nockerlStepVisualState(index, current, errorAt)
            val isLast = index == steps.lastIndex
            Row(modifier = Modifier.height(IntrinsicSize.Min)) {
                Column(
                    modifier = Modifier.width(STEP_DISC_SIZE),
                    horizontalAlignment = Alignment.CenterHorizontally,
                ) {
                    StepDisc(index, state)
                    if (!isLast) {
                        Box(
                            modifier =
                                Modifier
                                    .padding(vertical = 4.dp)
                                    .width(CONNECTOR_THICKNESS)
                                    .defaultMinSize(minHeight = 20.dp)
                                    .weight(1f)
                                    .clip(NockerlProgressTrackShape)
                                    .background(
                                        if (index < current) colors.accentPrimary else colors.divider,
                                    ),
                        )
                    }
                }
                Column(
                    modifier =
                        Modifier
                            .padding(start = 12.dp, bottom = if (isLast) 0.dp else 20.dp)
                            .stepClickable(state, index, onStepClick),
                    verticalArrangement = Arrangement.spacedBy(2.dp),
                ) {
                    StepLabel(step.label, state, alignCenter = false)
                    if (step.description != null) {
                        Text(
                            text = step.description,
                            style = MaterialTheme.typography.labelMedium,
                            color = colors.onCardMuted,
                        )
                    }
                }
            }
        }
    }
}

/** Disc + centered label, the horizontal step column (72dp keeps columns even). */
@Composable
private fun StepColumn(
    index: Int,
    step: NockerlStep,
    state: NockerlStepVisualState,
    onStepClick: ((Int) -> Unit)?,
) {
    Column(
        modifier = Modifier.width(HORIZONTAL_COLUMN_WIDTH).stepClickable(state, index, onStepClick),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        StepDisc(index, state)
        StepLabel(step.label, state, alignCenter = true)
    }
}

/** The 32dp step disc, per-state per the ratified recipe. */
@Composable
private fun StepDisc(
    index: Int,
    state: NockerlStepVisualState,
) {
    val colors = LocalNockerlColors.current
    val stateName =
        when (state) {
            NockerlStepVisualState.DONE -> "Completed"
            NockerlStepVisualState.CURRENT -> "Current step"
            NockerlStepVisualState.UPCOMING -> "Upcoming"
            NockerlStepVisualState.ERROR -> "Failed"
        }

    // The CURRENT disc wears the accentPrimarySoft halo ring (4dp).
    val halo =
        if (state == NockerlStepVisualState.CURRENT) {
            Modifier.background(colors.accentPrimarySoft, CircleShape).padding(RING_WIDTH)
        } else {
            Modifier.padding(RING_WIDTH)
        }

    Box(
        modifier = halo.semantics { stateDescription = stateName },
        contentAlignment = Alignment.Center,
    ) {
        when (state) {
            NockerlStepVisualState.DONE, NockerlStepVisualState.ERROR -> {
                val fill =
                    if (state == NockerlStepVisualState.DONE) colors.accentPrimary else colors.statusError
                Box(
                    modifier =
                        Modifier
                            .size(STEP_DISC_SIZE)
                            .nockerlShadow(elevation = NockerlElevation.Level1, shape = CircleShape)
                            .background(fill, CircleShape)
                            .nockerlLitSurface(shape = CircleShape),
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(
                        imageVector =
                            if (state == NockerlStepVisualState.DONE) Icons.Filled.Check else Icons.Filled.Close,
                        contentDescription = null, // named by stateDescription + label
                        modifier = Modifier.size(16.dp),
                        tint = pickOnAccent(fill),
                    )
                }
            }
            NockerlStepVisualState.CURRENT -> {
                Box(
                    modifier =
                        Modifier
                            .size(STEP_DISC_SIZE)
                            .nockerlShadow(elevation = NockerlElevation.Level1, shape = CircleShape)
                            .background(colors.accentPrimary, CircleShape)
                            .nockerlLitSurface(shape = CircleShape),
                    contentAlignment = Alignment.Center,
                ) {
                    StepNumber(index, pickOnAccent(colors.accentPrimary))
                }
            }
            NockerlStepVisualState.UPCOMING -> {
                Box(
                    modifier =
                        Modifier
                            .size(STEP_DISC_SIZE)
                            .nockerlRecessedSurface(shape = CircleShape)
                            .background(colors.canvasAlt)
                            .border(1.dp, colors.divider, CircleShape),
                    contentAlignment = Alignment.Center,
                ) {
                    StepNumber(index, colors.onCardMuted)
                }
            }
        }
    }
}

@Composable
private fun StepNumber(
    index: Int,
    color: androidx.compose.ui.graphics.Color,
) {
    Text(
        text = (index + 1).toString(),
        style = MaterialTheme.typography.bodyMedium,
        fontWeight = FontWeight.Medium, // 500 is the bold cap (law §11)
        color = color,
    )
}

@Composable
private fun StepLabel(
    label: String,
    state: NockerlStepVisualState,
    alignCenter: Boolean,
) {
    val colors = LocalNockerlColors.current
    Text(
        text = label,
        style = MaterialTheme.typography.labelMedium,
        fontWeight =
            if (state == NockerlStepVisualState.UPCOMING) FontWeight.Normal else FontWeight.Medium,
        color =
            when (state) {
                NockerlStepVisualState.ERROR -> colors.statusError
                NockerlStepVisualState.UPCOMING -> colors.onCardMuted
                else -> colors.onCard
            },
        textAlign = if (alignCenter) TextAlign.Center else TextAlign.Start,
    )
}

/** Horizontal connector segment. */
@Composable
private fun Connector(
    done: Boolean,
    modifier: Modifier,
) {
    val colors = LocalNockerlColors.current
    Box(
        modifier =
            modifier
                .height(CONNECTOR_THICKNESS)
                .clip(NockerlProgressTrackShape)
                .background(if (done) colors.accentPrimary else colors.divider),
    )
}

/** DONE steps become tappable when the host wires [onStepClick] (jump back only). */
private fun Modifier.stepClickable(
    state: NockerlStepVisualState,
    index: Int,
    onStepClick: ((Int) -> Unit)?,
): Modifier =
    if (onStepClick != null && state == NockerlStepVisualState.DONE) {
        clickable { onStepClick(index) }
    } else {
        this
    }

/** Disc diameter (32dp) and the ratified geometry constants. */
private val STEP_DISC_SIZE = 32.dp
private val CONNECTOR_THICKNESS = 3.dp
private val RING_WIDTH = 4.dp
private val HORIZONTAL_COLUMN_WIDTH = 72.dp

/** Vertical footprint: the `size.container.lg` token (360). */
private val VERTICAL_MAX_WIDTH = 360.dp
